endless-sky/endless-sky
# feat(conditional dialogs): `to display` in dialog text

**Feature:** Allows `to display` in dialog text, plus a few more features that came as a consequence of the code changes.

## Feature Details
Dialog text can have `to display`, all substitutions visible in conversations are expanded in dialog text at mission instantiation, and you'll get a warning if you have spurious "accept" lines in your dialogs.

~~I had to include #8503 in this so the three places in endless-sky master with spurious "accept" lines in dialogs won't cause the checks to fail.~~ (That PR has been merged.)

```ruby
mission "My Mission"
    ...
    on offer
        dialog
            # A ${phrase} is expanded, and any <substitution> available to a conversation.
            "Fly ${band} to <planet> for <payment>?"
            # Lines after the first will be indented by one tab even if they're missing a tab.
            "You're rich enough to handle pirates along the way."
                to display
                    credits > 10000000
            "Surely a few pirates won't deter you."
                to display
                    credits <= 10000000
                    "combat rating" > 100
            "I'm sure you'll survive the pirates."
                to display
                    credits <= 10000000
                    "combat rating" <= 100
            # Stock phrases can have "to display"
            phrase "hostile disabled"
                to display
                    "combat rating" > 10000
            # Anonymous phrases can NOT have "to display"
            # That is because it would require making large changes to the Phrase class.
            # Changes where there is no clear best way to implement them.
            phrase
                word
                    "Frelling pidgeon-lover." 100
                    "This is my least favorite phrase." 1
```

## UI Screenshots
N/A

## Usage Examples
1. Dialog text depends on player's actions or choices elsewhere in the story.
2. Substitutions that were not available before are now, which means some things that had to be conversations can be dialogs now.
3. Put multiple `${phrases}` in a dialog without having to create a new phrase to do it.

## Testing Done
1. Copy this into data/[dialogtest.txt](https://github.com/user-attachments/files/15522834/dialogtest.txt)
2. Fly to Earth and visit the spaceport for instructions.
3. Fly to Mars to pick 1 or 2 in a conversation
4. Fly to Luna to see the correct entry show up in a dialog.
5. Repeat steps 3 and 4 ad nauseam.

### Automated Tests Added
N/A

## Performance Impact
Negligible.
